lib/output.rb

```diff
@@ -2,106 +2,124 @@
 require_relative './colors'
 require_relative './secret_code'
 
-class Output
-
-  class << self
-    TITLE_STARS = (Colors.yellow_b('*******************************')).center(85)
-    TITLE = (Colors.yellow_b('*****') + Colors.red_b('Welcome to Mastermind') + Colors.yellow_b('*****')).center(103)
-    RULES = """
-      The rules are simple, the Computer will generate a secret code
-    that you will have to guess. There are 6 colors with the code containing
-    4 colors in which the order matters.
-
-      You will input your guesses in the form of \"[color] [color] [color]
-    [color]\". In which you have 6 color options: Red('r'), Orange('o'),
-    Yellow('y'), Green('g'), Blue('b'), Purple('p'). (Atm your input will be
-    taken in numbers instead with the numbers 1-6 referring to the colors in
-    the respective order previously mentioned)
-
-      The computer will give you feedback in the form of \"[peg type] [peg
-    type] [peg type] [peg type]\" in which the peg type will either be white,
-    red, or none:
-        White: Your guess is the right color, but wrong position.
-        Red: Your guess is at the right color and the right position.
-        None: This color is not part of the secret code.
-
-      You win if you are able to guess the code within 10 tries, otherwise
-    the computer wins. You can quit the game after anytime that you receive
-    feedback by typing 'quit'.
-    """
-    # INPUT_PROMPT = "\nYou have " + Colors.cyan(guesses_left_string) + ' guesses left' + "\nWhat is your guess?\n\t" + Colors.purple_b('>>>')
-    GUESSES_LEFT = "\nGuesses left: "
-    INPUT_PROMPT = "What is your guess?\n\t" + Colors.purple_b('>>>')
-    INPUT_ERROR = Colors.red_b("\nSorry, that's not the correct input:\n\tPlease refer to the instructions for the input method")
-    CODE_GENERATED = Colors.blue_b('The secret code has been generated, you may begin . . .')
-    GAME_BEGIN = Colors.green_b('Would you like to begin? [y/n]')
-    GAME_AGAIN = Colors.green_b('Would you like to go again? [y/n]')
-    NO_GUESSES = Colors.red_b('Nice try, but you ran out of guesses :(')
-    THE_ANSWER = Colors.white_b('The secret code was: ')
-
-    def print_intro
-      puts ''
-      header
-      instructions
-      type_ok
-      system 'clear'
-    end
+module Output
+  include Colors
```

Comment:
A really cool thing about modules is that they can share behavior with a given object, but it's important to be diligent in it's use to avoid violating the [Liskov Substitution Principle](https://www.tomdalling.com/blog/software-design/solid-class-design-the-liskov-substitution-principle/).

I like the *is/behaves like/has* smell test, it helps decide how to include external behavior into a class. The most coherent phrasing would (respectively) indicate either to use inheritance, mix-ins, or dependency injection. If none of the sentences make sense, I would create a higher level abstraction to coordinate between the namespaces.

So in this situation, our three indicators would be
* output _is_ colors
* output _behaves like_ colors
* output _has_ colors

While _output behaves like_ colors sounds OK, I think that _output has colors_ makes the most sense in this given situation, so IMO both `Output` and `Colors` might be best suited as classes, with `Colors` as a dependency of the `Output`.

Here's some other semi-related topics to look into:
[Composition over Inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)
[Modules](http://ruby-doc.com/docs/ProgrammingRuby/html/tut_modules.html)
[Why Mixins are Broken](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html#why-mixins-are-broken)

That last shared article talks specifically about reactjs, but the concepts under the "Why Mixins are Broken" header apply to ruby mixins just as well.

Read some of that over, and let me know what you think.

Reply:
Ok so from the articles you linked and some other stuff I read Mix-ins are defining methods to be used over potentially multiple classes, but Inheritance is more so laying out an empty or default method that can be overridden over multiple classes.

So in this case you recommend to use neither of those, but instead dependency injection where I would simply create a Color object within Output and have it run it's methods when it is necessary. Would that be the correct way of implementing dependency injection in this case?